hail/src/main/scala/is/hail/expr/ir/ndarrays/EmitNDArray.scala

```diff
@@ -275,56 +275,51 @@ object EmitNDArray {
-                  cb.assign(missing, false)
-                  // Need to check if the any of the ndarrays are missing.
-                  val missingCheckLoopIdx = cb.newLocal[Int]("ndarray_concat_missing_check_idx")
-                  cb.forLoop(cb.assign(missingCheckLoopIdx, 0), missingCheckLoopIdx < arrLength, cb.assign(missingCheckLoopIdx, missingCheckLoopIdx + 1),
-                    cb.assign(missing, missing | ndsArraySValue.isElementMissing(cb, missingCheckLoopIdx))
-                  )
-                  missing
-                }
-              }
+              IEmitCode(cb, ndsArraySValue.hasMissingValues(cb), {
+                val firstND = ndsArraySValue.loadElement(cb, 0).get(cb).asNDArray
 
-              IEmitCode(cb, missing, {
-                val loopIdx = cb.newLocal[Int]("ndarray_concat_shape_check_idx")
-                val firstND = ndsArraySValue.loadElement(cb, 0).map(cb) { sCode => sCode.asNDArray }.get(cb)
+                // compute array of sizes along concat axis, and total size of concat axis
+                val arrayLongPType = PCanonicalArray(PInt64(), true)
+                val (pushSize, finishSizes) = arrayLongPType.constructFromFunctions(cb, region, arrLength, false)
 
-                val stagedArrayOfSizesPType = PCanonicalArray(PInt64(), true)
-                val (pushElement, finish) = stagedArrayOfSizesPType.constructFromFunctions(cb, region, arrLength, false)
+                val concatAxisSize = cb.newLocal[Long](s"ndarray_concat_axis_size", 0L)
 
-                val newShape = (0 until x.typ.nDims).map { dimIdx =>
-                  val localDim = cb.newLocal[Long](s"ndarray_concat_output_shape_element_${dimIdx}")
-                  val ndShape = firstND.shapes
-                  cb.assign(localDim, ndShape(dimIdx))
-                  if (dimIdx == axis) {
-                    pushElement(cb, EmitCode(Code._empty, false, primitive(localDim)).toI(cb))
-                  }
+                ndsArraySValue.forEachDefined(cb) { (cb, i, nd) =>
+                  val dimLength = nd.asNDArray.shapes(axis)
+                  pushSize(cb, EmitCode(Code._empty, false, primitive(dimLength)).toI(cb))
+                  cb.assign(concatAxisSize, concatAxisSize + dimLength)
+                }
+                val stagedArrayOfSizes = finishSizes(cb)
 
-                  cb.forLoop(cb.assign(loopIdx, 1), loopIdx < arrLength, cb.assign(loopIdx, loopIdx + 1), {
-                    val shapeOfNDAtIdx = ndsArraySValue.loadElement(cb, loopIdx).map(cb) { sCode => sCode.asNDArray }.get(cb).shapeStruct(cb)
-                    val dimLength = cb.newLocal[Long]("dimLength", shapeOfNDAtIdx.loadField(cb, dimIdx).get(cb).asInt64.value)
+                // compute index of first input which has non-zero concat axis size
+                val firstNonEmpty = cb.newLocal[Int]("ndarray_concat_first_nonempty", 0)
+                cb.whileLoop(stagedArrayOfSizes.loadElement(cb, firstNonEmpty).get(cb).asInt64.value.ceq(0L), {
```

Comment:
is this checking is null? don't we have a `isna` function to make this clearer?

Reply:
No, this is checking for size 0 axes.

Reply:
Oh yes, arrayOfSizes makes this kinda obvious. duh.